[feature][card] Add API v2 Card type support

woohgit opened the pull request:
I love the new v2 API supported card type messages.

For now the messages would look like these below.

Feel free to test it and let me know if you like it. I think this card type can merge the short and extended notification, since you can check the details by expanding the message. We just need to define what details should go on the collapsed line and what details on the expanded box.

Any comments are welcome.

## all collapsed by default
![all](http://i.imgur.com/WjQVYT6.png)

## services expanded
![services expanded](http://i.imgur.com/ay0ZX5S.png)

## hosts expanded
![hosts expanded](http://i.imgur.com/umbNOdE.png)

woohgit committed:
[feature][card] Add card support

woohgit committed:
[feature][card] Remove unused code

asdf1234-deleted commented:
This would be a great addition!

asdf1234-deleted commented:
How can this be tested?

```
hipsaint-test --host HIPCHAT_HOST --user some_user --token SOME_TOKEN --room 'operations-testing' --type host --inputs "$HOSTNAME|$LONGDATETIME|$NOTIFICATIONTYPE|$HOSTADDRESS|$HOSTSTATE|$HOSTOUTPUT" -n -V 2 --card
```


```
  File "/usr/lib64/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib64/python2.7/urllib2.py", line 437, in open
    response = meth(req, response)
  File "/usr/lib64/python2.7/urllib2.py", line 550, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python2.7/urllib2.py", line 475, in error
    return self._call_chain(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 558, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 400: Bad Request
```

woohgit commented:
@mtdeguzis I Use them as follows:

```
define command {
    command_name    notify-host-by-hipchat
    command_line    /opt/aws/bin/hipsaint -V 2 --token={{ lookup('vault', 'secret/production/servers/prd-nagios-ops-01', 'hipchat_token') }} --room={{ lookup('vault', 'secret/production/servers/prd-nagios-ops-01', 'hipchat_room') }} --type=host --inputs="$HOSTNAME$|$LONGDATETIME$|$NOTIFICATIONTYPE$|$HOSTADDRESS$|$HOSTSTATE$|$HOSTOUTPUT$" -n -c
}
define command {
    command_name    notify-host-by-hipchat-room
    command_line    /opt/aws/bin/hipsaint -V 2 --room=$ARG1$ --token=$ARG2$ --type=host --inputs="$HOSTNAME$|$LONGDATETIME$|$NOTIFICATIONTYPE$|$HOSTADDRESS$|$HOSTSTATE$|$HOSTOUTPUT$" -n -c
}
define command {
    command_name    notify-service-by-hipchat
    command_line    /opt/aws/bin/hipsaint -V 2 --token={{ lookup('vault', 'secret/production/servers/prd-nagios-ops-01', 'hipchat_token') }} --room={{ lookup('vault', 'secret/production/servers/prd-nagios-ops-01', 'hipchat_room') }} --type=short-service --inputs="$SERVICEDESC$|$HOSTALIAS$|$LONGDATETIME$|$NOTIFICATIONTYPE$|$HOSTADDRESS$|$SERVICESTATE$|$SERVICEOUTPUT$" -n -c
}
define command {
    command_name    notify-service-by-hipchat-room
    command_line    /opt/aws/bin/hipsaint -V 2 --room=$ARG1$ --token=$ARG2$ --type=short-service --inputs="$SERVICEDESC$|$HOSTALIAS$|$LONGDATETIME$|$NOTIFICATIONTYPE$|$HOSTADDRESS$|$SERVICESTATE$|$SERVICEOUTPUT$" -n -c
}
```

mtdeguzis commented:
Ok, I was adding --card, per the commit, but it works with that left out.

asdf1234-deleted commented:
That you for your work! We really appreciate this here where I work. I noticed you can't use short-host, but have to use host, or an argument parser exception shows up. I am curious on how to change the icon (ours is blank) but I may be able to figure that out.

asdf1234-deleted commented:
Ok I see:

```
ICON_URL = 'http://i.imgur.com/BxqWsDC.png'
```

asdf1234-deleted committed:
Ensure long output does not exceed API limits

With some of our simple checks, long output cause an issue:

```
Traceback (most recent call last):
  File "/home/icinga/.local/share/virtualenvs/icinga-plugins--BT7bI7t/bin/hipsaint", line 11, in <module>
    load_entry_point('hipsaint', 'console_scripts', 'hipsaint')()
  File "/opt/icinga-plugins/hipsaint-test/hipsaint/bin/commands.py", line 85, in main
    msg.deliver_payload()
  File "/opt/icinga-plugins/hipsaint-test/hipsaint/messages.py", line 102, in deliver_payload_v2
    raw_response = urlopen(request)
  File "/usr/lib64/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib64/python2.7/urllib2.py", line 437, in open
    response = meth(req, response)
  File "/usr/lib64/python2.7/urllib2.py", line 550, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib64/python2.7/urllib2.py", line 475, in error
    return self._call_chain(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 409, in _call_chain
    result = func(*args)
  File "/usr/lib64/python2.7/urllib2.py", line 558, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
urllib2.HTTPError: HTTP Error 400: Bad Request
```

This is due to limitations set by the hipchat API. This proposed change detects and truncates long output for the card. It does not affect what is seen in an email or in Icingaweb2.

woohgit committed:
Merge pull request #1 from mtdeguzis/patch-1

Ensure long output does not exceed API limits

woohgit committed:
Fix formatting

woohgit commented:
No more HipChat.
